app/uk/gov/hmrc/hec/controllers/IFController.scala

```diff
@@ -0,0 +1,112 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.hec.controllers
+
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+import com.google.inject.{Inject, Singleton}
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import uk.gov.hmrc.hec.models.TaxYear
+import uk.gov.hmrc.hec.models.ids.{CTUTR, SAUTR}
+import uk.gov.hmrc.hec.services.IFService
+import uk.gov.hmrc.hec.util.Logging
+import uk.gov.hmrc.hec.util.Logging.LoggerOps
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import cats.data.Validated
+import cats.data.Validated._
+import cats.implicits._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class IFController @Inject() (
+  IFService: IFService,
+  cc: ControllerComponents
+)(implicit ec: ExecutionContext)
+    extends BackendController(cc)
+    with Logging {
+
+  /**
+    * Fetch individual user's self-assessment status
+    * @param utr Self-assessment UTR
+    * @param taxYear Start tax year
+    * @return Self-assessment status with UTR and tax year
+    */
+  def getSAStatus(utr: String, taxYear: String): Action[AnyContent] = Action.async { implicit request =>
+    val sautrValidation   = SAUTR.fromString(utr).toValid("Invalid SAUTR")
+    val taxYearValidation = TaxYear.fromString(taxYear).toValid("Invalid tax year")
+
+    @SuppressWarnings(Array("org.wartremover.warts.Any"))
+    val validation = (sautrValidation, taxYearValidation).mapN((_, _))
+    validation match {
+      case Valid((utr, year)) =>
+        IFService
+          .getSAStatus(utr, year)
+          .fold(
+            { e =>
+              logger.warn("Could not fetch SA status", e)
+              InternalServerError
+            },
+            saStatus => Ok(Json.toJson(saStatus))
+          )
+      case Invalid(e)         => Future.successful(BadRequest(e.toList.mkString(";")))
+    }
+  }
+
+  /**
+    * Fetch a company's corporation tax status
+    * @param utr The corporation tax UTR
+    * @param fromDate The start date of the tax period
+    * @param toDate The end date of the tax period
+    * @return Corporation tax status & accounting periods
+    */
+  def getCTStatus(
+    utr: String,
+    fromDate: String,
+    toDate: String
+  ): Action[AnyContent] = Action.async { implicit request =>
+    def parsedDate(dateStr: String, error: String): Validated[String, LocalDate] =
+      try LocalDate.parse(dateStr, DateTimeFormatter.ISO_DATE).valid
+      catch {
+        case _: Exception => invalid(error)
+      }
+
+    val ctutrValidation: Validated[String, CTUTR]        = CTUTR.fromString(utr).toValid("Invalid CTUTR")
+    val fromDateValidation: Validated[String, LocalDate] = parsedDate(fromDate, "Invalid fromDate format")
+    val toDateValidation: Validated[String, LocalDate]   = parsedDate(toDate, "Invalid toDate format")
```

Comment:
I think you may need to do `toValidNel` on the `Options` above and make `parsedDate` return a `ValidatedNel[String, LocalDate]`. Without the `NonEmptyList` on the error side, I'm not sure how `mapN` will accumlate the errors - it could just concatenate all the strings or just select one

Reply:
yeah, I just noticed that it was concatenating the error strings, will update 👍